### ISSUE TITLE
ArrayMesh ARRAY_COLOR values are clamped to [0, 1] range when passed to shader.

### ISSUE THREAD

**Issue by Giwayume** [godotengine/godot#66197]
### Godot version

3.5.stable

### System information

Windows 7 64 bit

### Issue description

Vertex colors in ArrayMesh definition are being clipped to a [0,1] range unnecessarily, severely limiting the value of using vertex color for a non-traditional means of passing interpolated data to the fragment shader.

### Steps to reproduce

Constructing a PoolColorArray() with values outside the 2.0 range (in this case red).

```
var vertices = PoolVector3Array()
vertices.push_back(Vector3(0, 1, 0) * 300)
vertices.push_back(Vector3(1, 0, 0) * 300)
vertices.push_back(Vector3(0, 0, 1) * 300)

var colors = PoolColorArray()
colors.push_back(Color(2.0, 0.0, 0.0, 1.0))
colors.push_back(Color(2.0, 0.0, 0.0, 1.0))
colors.push_back(Color(2.0, 0.0, 0.0, 1.0))

# Initialize the ArrayMesh.
var arr_mesh = ArrayMesh.new()
var arrays = []
arrays.resize(ArrayMesh.ARRAY_MAX)
arrays[ArrayMesh.ARRAY_VERTEX] = vertices
arrays[ArrayMesh.ARRAY_COLOR] = colors
# Create the Mesh.
arr_mesh.add_surface_from_arrays(Mesh.PRIMITIVE_TRIANGLES, arrays)
var m = MeshInstance.new()
m.mesh = arr_mesh
```

Once the vertex color reaches the shader, the color is clipped to 0-1 range. I know because this shader code darkens the red.

```
shader_type canvas_item;

void vertex() {
	COLOR = vec4(COLOR.r / 2.0, COLOR.gba);
}
```

Polygon2D does not have this same limitation. When you define vertex colors there, their values are not clipped in the shader.

**Comment by lawnjelly** [comment]
You don't mention GLES2 or 3, but in GLES3 the code is here:

```
			case VS::ARRAY_COLOR: {
				attribs[i].size = 4;

				if (p_format & VS::ARRAY_COMPRESS_COLOR) {
					attribs[i].type = GL_UNSIGNED_BYTE;
					attributes_stride += 4;
					attribs[i].normalized = GL_TRUE;
				} else {
					attribs[i].type = GL_FLOAT;
					attributes_stride += 16;
					attribs[i].normalized = GL_FALSE;
				}

			} break;
```
and
```
		ARRAY_COMPRESS_DEFAULT = ARRAY_COMPRESS_NORMAL | ARRAY_COMPRESS_TANGENT | ARRAY_COMPRESS_COLOR | ARRAY_COMPRESS_TEX_UV | ARRAY_COMPRESS_TEX_UV2 | ARRAY_COMPRESS_WEIGHTS | ARRAY_FLAG_USE_OCTAHEDRAL_COMPRESSION
```
So I'm guessing the color data is compressed hence being normalized.

I'm not sure if you can override the compression from the user side I will try and work it out.

**Comment by Calinou** [comment]
Unlike 3.x, mesh vertex colors are now RGB8 in 4.0 for performance reasons. If you need higher precision, use the custom vertex data fields which are new in 4.0.

**Comment by Giwayume** [comment]
Thanks, I did not notice the compression option when adding the surface. And it's good to know there's a workaround in 4.0.

@Calinou are the custom vertex data fields interpolated across the face of the mesh like UV or vertex colors? That's the part I'm relying on.

**Comment by Calinou** [comment]
> are the custom vertex data fields interpolated across the face of the mesh like UV or vertex colors? That's the part I'm relying on.

I think they are, unless you can use a `flat` interpolation specifier somehow.

**Comment by akien-mga** [comment]
Fixed by #66225.

**Comment by Giwayume** [comment]
Ok, I thought everything would be well with Godot 4, however, custom data isn't passed to canvas item shaders!

So basically, I am more limited in Godot 4 than I am in Godot 3, where I can utilize vertex color as uncompressed interpolated float values in 2D. I'm not seeing an alternative.
